languages/rust/src/query.rs

```diff
@@ -0,0 +1,282 @@
+    fn next(&mut self) -> Option<Self::Item> {
+        Query::next_result(self)
+    }
+}
+
+pub struct Query {
+    inner: polar_core::polar::Query,
+    calls: HashMap<u64, Box<dyn Iterator<Item = Arc<dyn crate::host::ToPolar>>>>,
+    host: Arc<Mutex<crate::host::Host>>,
+}
+
+impl Query {
+    pub fn new(inner: polar_core::polar::Query, host: Arc<Mutex<crate::host::Host>>) -> Self {
+        Self {
+            calls: HashMap::new(),
+            inner,
+            host,
+        }
+    }
+
+    pub fn next_result(&mut self) -> Option<anyhow::Result<ResultSet>> {
+        loop {
+            let event = self.inner.next()?;
+            self.check_messages();
+            if let Err(e) = event {
+                return Some(Err(e.into()));
+            }
+            let event = event.unwrap();
+            tracing::debug!(event=?event);
+            let result = match event {
+                QueryEvent::None => Ok(()),
+                QueryEvent::Done => return None,
+                QueryEvent::Result { bindings, .. } => {
+                    return Some(Ok(ResultSet {
+                        bindings,
+                        host: self.host.clone(),
+                    }));
+                }
+                QueryEvent::MakeExternal {
+                    instance_id,
+                    constructor,
+                } => self.handle_make_external(instance_id, constructor),
+                QueryEvent::ExternalCall {
+                    call_id,
+                    instance,
+                    attribute,
+                    args,
+                } => self.handle_external_call(call_id, instance, attribute, args),
+                QueryEvent::ExternalOp {
+                    call_id,
+                    operator,
+                    args,
+                } => self.handle_external_op(call_id, operator, args),
+                QueryEvent::ExternalIsa {
+                    call_id,
+                    instance,
+                    class_tag,
+                } => self.handle_external_isa(call_id, instance, class_tag),
+                QueryEvent::ExternalUnify {
+                    call_id,
+                    left_instance_id,
+                    right_instance_id,
+                } => self.handle_external_unify(call_id, left_instance_id, right_instance_id),
+                QueryEvent::ExternalIsSubSpecializer {
+                    call_id,
+                    instance_id,
+                    left_class_tag,
+                    right_class_tag,
+                } => self.handle_external_is_subspecializer(
+                    call_id,
+                    instance_id,
+                    left_class_tag,
+                    right_class_tag,
+                ),
+                QueryEvent::Debug { message } => self.handle_debug(message),
+            };
+            if let Err(e) = result {
+                self.application_error(e);
+            }
+        }
+    }
+
+    fn check_messages(&mut self) {
```

Comment:
This duplicates the implementation on `Polar`.  Can we just have one?

Reply:
Probably, yes.